docs/proposals/external-lb-source-ip-preservation.md

```diff
@@ -0,0 +1,240 @@
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN MUNGE: GENERATED_TOC -->
+
+- [Overview](#overview)
+  - [Motivation](#motivation)
+  - [Related work](#related-work)
+- [Alpha Design](#alpha-design)
+  - [Overview](#overview-1)
+  - [Prerequisites](#prerequisites)
+  - [API Changes](#api-changes)
+    - [Pod Security Policy](#pod-security-policy)
+  - [Deploying profiles](#deploying-profiles)
+  - [Testing](#testing)
+- [Beta Design](#beta-design)
+  - [API Changes](#api-changes-1)
+- [Future work](#future-work)
+  - [System component profiles](#system-component-profiles)
+  - [Deploying profiles](#deploying-profiles-1)
+  - [Custom app profiles](#custom-app-profiles)
+  - [Security plugins](#security-plugins)
+  - [Container Runtime Interface](#container-runtime-interface)
+  - [Alerting](#alerting)
+  - [Profile authoring](#profile-authoring)
+- [Appendix](#appendix)
+
+<!-- END MUNGE: GENERATED_TOC -->
+
+# Overview
+
+Kubernetes provides an external loadbalancer service type which creates a virtual external ip
+(in supported cloud provider environments) that can be used to load-balance traffic to
+the pods matching the service pod-selector.
+
+## Motivation
+
+The current implemention requires that the cloud loadbalancer balances traffic across all
+Kubernetes worker nodes, and this traffic is then equally distributed to all the backend
+pods for that service.
+Due to the DNAT required to redirect the traffic to its ultimate destination, the return
+path for each session MUST traverse the same node again. To ensure this, the node also
+performs a SNAT, replacing the source ip with its own.
+
+This causes the service endpoint to see the session as originating from a cluster local ip address.
+*The original external source IP is lost*
+
+This is not a satisfactory solution - the original external source IP MUST be preserved for a
+lot of applications and customer use-cases.
+
+# Alpha Design
+
+This section describes the proposed design for
+[alpha-level](../../docs/devel/api_changes.md#alpha-beta-and-stable-versions) support, although
+additional features are described in [future work](#future-work).
+
+## Overview
+
+The double hop must be prevented by programming the external load balancer to direct traffic
+only to nodes that have local pods for the service. This can be accomplished in two ways, either
+by API calls to add/delete nodes from the LB node pool or by adding health checking to the LB and
+failing/passing health checks depending on the presence of local pods.
+
```

Comment:
Yes, we can only go down a certain path since the packet paths are already implemented.
Unfortunately, we don't have a detailed "life-of-a-packet" overview for the existing path and fixups, but for this change, I will try to convert some slides into SVGs - its hard to embed images into Markdown docs that render correctly everywhere in Github and locally.

In GCE, there is no DNAT performed, the packets arrive on the GCE VMs with the Virtual IP still intact.